File: Difficulty/Easy/[100] Same Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        qp, qq = collections.deque(), collections.deque()
        if p is None and q is None:
            return True
        if p is None or q is None:
            return False
        qp.append(p)
        qq.append(q)
        while qp and qq:
            p = qp.popleft()
            q = qq.popleft()
            if p.val != q.val:
                return False
            if p.left is not None and q.left is not None:
                qp.append(p.left)
                qq.append(q.left)
            elif p.left is None and q.left is None:
                pass
            else:
                return False
            if p.right is not None and q.right is not None:
                qp.append(p.right)
                qq.append(q.right)
            elif p.right is None and q.right is None:
                pass
            else:
                return False
        return True

File: Difficulty/Easy/[100] Same Tree/solution.cpp
/*
 * @lc app=leetcode id=100 lang=cpp
 *
 * [100] Same Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <queue>

class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        queue<TreeNode*> qp, qq;
        if (p == nullptr && q == nullptr) return true;
        if (p == nullptr || q == nullptr) return false;
        qp.push(p);
        qq.push(q);
        while(!qp.empty() && !qq.empty()) {
            p = qp.front();
            q = qq.front();
            if (p->val != q->val)
                return false;

            if (p->left != nullptr && q->left != nullptr){
                qp.push(p->left);
                qq.push(q->left);
            }else if (p->left == nullptr && q->left == nullptr){}
            else return false;

            if (p->right != nullptr && q->right != nullptr){
                qp.push(p->right);
                qq.push(q->right);
            }else if (p->right == nullptr && q->right == nullptr){}
            else return false;
            
            qp.pop();
            qq.pop();
        }
        return true;
    }
};
// @lc code=end

